### PR TITLE
restart on traefik-route initiated static config changes

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds a 'static' field to the traefik-route interface and instructs Traefik to take whatever's in there and put it into the static config and restart itself whenever there's a change.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -706,18 +718,49 @@ def _process_ingress_relation(self, relation: Relation):
...
+
+    def _traefik_route_static_configs(self):
+        """Fetch all static configurations passed through traefik route."""
+        configs = []
+        for relation in self.traefik_route.relations:
+            config = self.traefik_route.get_static_config(relation)
+            if config:
+                dct = self._try_load_dict(config)
+                if not dct:
+                    continue
```
Comment (mmkay) [inline src/charm.py:742]: Log and continue, maybe?

**Review by mmkay — COMMENTED** [review]
First round done, two things come up. Will see it again once CI is happy.

File: src/traefik.py [hunk 2]
```diff
@@ -241,8 +266,24 @@ def _update_static_configuration(self):
...
+        # we attempt to put together the base config with whatever the user passed via traefik_route.
+        # in case there are conflicts between the base config and some route, or between the routes themselves,
+        # we'll be forced to bail out.
+        additional_static_configs = list(self._traefik_route_static_configs)
+        if additional_static_configs:
+            static_configs = [base_config] + additional_static_configs
+            try:
+                return reduce(static_config_deep_merge, static_configs)
+            except StaticConfigMergeConflictError as e:
+                logger.exception("failed to merge static config into traefik's base config."
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply src/traefik.py:279]: it's not that easy to surface this as a status. We'd need @sed-i's push-status API.

[the remark this answers is not in the document]
Comment (sed-i) [reply src/traefik.py:279]: Config issue here seems like a major issue. Could charm code let traefik keep running with broken config? What would happen if a different lifecycle event comes in?

My guess is that you would need a mechanism that always keeps the charm in Blocked until this is revolved. One way to achieve this is an exception that is caught by the caller, and then use a status dict + return. [Example](https://github.com/canonical/prometheus-k8s-operator/blob/6eea5cc49bce622c089c54a7be0f891e67966c5c/src/charm.py#L538).

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply src/traefik.py]: if there is any issue with the static config the current behaviour is to ignore the broken parts (those coming from traefik-route) and configure traefik with the 'baseline' static config plus the nonconflicting configs.

So it's not going to **break** any other route or traefik in any meaningful way.
Only, it's not going to be applied. 
Traefik will display a short-lived blocked status (will be overwritten by any other event) and will keep firing error logs any time it tries to drop the static config.

I agree it's not enough but refactoring traefik's status management is very out of scope for this PR and the problems go much deeper than traefik-route's management.

Opened #329 to track.

**Review by lucabello — APPROVED** [review]

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 3]
```diff
@@ -344,11 +356,14 @@ def is_ready(self) -> bool:
...
 
-    def submit_to_traefik(self, config):
+    def submit_to_traefik(self, config: dict, static: dict=None):
         """Relay an ingress configuration data structure to traefik.
 
         This will publish to TraefikRoute's traefik-route relation databag
         the config traefik needs to route the units behind this charm.
+
+        Use ``static`` if you need to update traefik's **static** configuration.
+        Note that this will force traefik to restart to comply.
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:367]: good point, made it more clear.

**Comment by PietroPasotti** [comment]
CI is stuck because of https://github.com/microsoft/linux-package-repositories/issues/130 now trying to run them locally, if they pass I vote we merge on red

**Comment by PietroPasotti** [comment]
only test failing locally is unrelated to these changes. (it's on forward-auth)

**Review by mmkay — APPROVED** [review]
Screenshot from 2024-04-24 14-56-58

Looks like it does what it was supposed to do :) Nice!

**Pull request merged**